### ISSUE TITLE
How to handle ConcurrentHashMap.keySet() signature change of Java 8?

### ISSUE THREAD

**Issue by ssikuro** [stefan-zobel/streamsupport#9]
I have a piece of code written for Java 8 that uses `java.util.concurrent.ConcurrentHashMap` and its `keySet()` method. I could use retrolambda to downgrade the class files to the Java-7 byte code successfully.

But at run time, I get:
```java.lang.NoSuchMethodError: java.util.concurrent.ConcurrentHashMap.keySet()Ljava/util/concurrent/ConcurrentHashMap$KeySetView;```

I think this is because the `keySet()` method of Java 7 returns the interface `Set<K>` while the same method returns the inner class `ConcurrentHashMap.KeySetView<K,V>`, which implements Set<K>. Because the code is compiled with JDK 8, it uses the method signature that returns `KeySetView<K,V>`. But at run time, the Java 7 JDK library only has the `keySet()` that returns `Set<K>`, and this code results in the `NoSuchMethodError`.

I see StreamSupport has CHM.java. The name sounds like from ConcurrentHashMap but it's under the alternative directory.  Does this mean it's an Android variant of ConcurrentHashMap and not a direct replacement?

Do you have any suggestion on how to handle this situation?

**Comment by stefan-zobel** [comment]
Hi Kuro,

if I understand you correctly, you have some compiled Java 8 code (Bytecode version 52) that you've run through retrolambda to get Java 7 bytecode.
And that Java 8 bytecode contains references to the Java 8 `KeySetView`.

Sounds familiar. I had that same [issue](https://sourceforge.net/p/streamsupport/tickets/3/) a couple of years ago in my own code (streamsupport).
But since I had control over the source code it was relatively straightforward to fix.

Without access to the source code it might turn out to be impossible. Perhaps [ProGuard](https://github.com/Guardsquare/proguard) is worth trying (instead or in addition to retrolambda). It has some Java 8 backporting capabilities, perhaps it can handle this situation (though I doubt that).

The `CHM` you've found was indeed intended as a (internal use only) emergency replacement for the Android ConcurrentHashMap which had a serious concurrency bug several years ago.
In the end I've never used it and I see no way you could take advantage of it.

To sum up: should be easy if you have access to the source code, requires bytecode rewriting if you don't (hard).

HTH,
Stefan

**Comment by ssikuro** [comment]
Stefan,
Thank you for getting back to me quickly. I was hoping this is a common problem and you have a magic solution but I guess it isn't that popular.
The root cause of this is that Oracle decided the return type of keySet() from a generic interface to a new generic class that implements the interface. While it works most cases, it won't work for the strategy of compiling with Java 8 and downgrading byte code to Java 7. They shouldn't have done that.
I do have the source code. I guess I have to change it to avoid using the `keySet()` method.

**Comment by stefan-zobel** [comment]
Hi Kuro,

your analysis is 100% correct. If you have access to the source it is probably enough to change the static type of the `map` instance on which you are calling `keySet()`.
That did it in my case at least. If I remember correctly, I changed from `ConcurrentHashMap` to `ConcurrentMap` and the reference to the `KeySetView` disappeared.

Regards,
Stefan

**Comment by ssikuro** [comment]
Thank you for sharing your solution.
In my case, I was able to solve the issue by assigning the ConcurrentHashMap instance to a Map variable, where `keySet()` is just `Map<K>`. Luckily the code was not using any method that is only available to ConcurrentHashMap.
